vllm/v1/core/scheduler.py

```diff
@@ -60,36 +65,61 @@ def __init__(
+        # Encoder-related.
+        # NOTE(woosuk): Here, "encoder" includes the vision encoder. Currently,
+        # we assume that the encoder also has the Transformer architecture
+        # (e.g., ViT).
+        # FIXME(woosuk): Below are placeholder values. We need to calculate the
+        # actual values from the configurations.
+        self.max_num_encoder_input_tokens = 2048
+        self.encoder_cache_manager = EncoderCacheManager(cache_size=2048)
 
+    def schedule(self) -> "SchedulerOutput":
         # NOTE(woosuk) on the scheduling algorithm:
         # There's no "decoding phase" nor "prefill phase" in the scheduler.
         # Each request just has the num_computed_tokens and num_tokens,
         # which is equal to len(prompt_token_ids) + len(output_token_ids).
         # At each step, the scheduler tries to assign tokens to the requests
         # so that each request's num_computed_tokens can catch up its
         # num_tokens. This is general enough to cover chunked prefills,
-        # prefix caching, and the "jump forward" optimization in the future.
+        # prefix caching, and the "jump decoding" optimization in the future.
+
+        scheduled_new_reqs: List[Request] = []
+        scheduled_resumed_reqs: List[Request] = []
+        scheduled_running_reqs: List[Request] = []
+        preempted_reqs: List[Request] = []
 
         req_to_new_block_ids: Dict[str, List[int]] = {}
         num_scheduled_tokens: Dict[str, int] = {}
         token_budget = self.max_num_scheduled_tokens
+        # Encoder-related.
+        scheduled_encoder_inputs: Dict[str, List[int]] = {}
+        encoder_budget = self.max_num_encoder_input_tokens
 
         # First, schedule the RUNNING requests.
+        # NOTE(woosuk): At most 1 request in the RUNNING queue is allowed to be
+        # in the "partial" state, where the request has some tokens computed
+        # but not all. The constraint is due to the persistent batch in the
+        # V1 model runner.
+        # TODO(woosuk): Remove this constraint after refactoring model runner.
+        has_partial_request = False
         req_index = 0
         while req_index < len(self.running):
-            if token_budget == 0:
-                break
-
+            # Only the last request in the RUNNING queue can be "partial".
+            assert not has_partial_request
+            assert token_budget > 0
             request = self.running[req_index]
             num_new_tokens = request.num_tokens - request.num_computed_tokens
             num_new_tokens = min(num_new_tokens, token_budget)
             assert num_new_tokens > 0
 
+            # Schedule encoder inputs.
+            encoder_inputs_to_schedule, num_new_tokens = (
+                self._schedule_encoder_inputs(request,
+                                              request.num_computed_tokens,
+                                              num_new_tokens, encoder_budget))
+            assert num_new_tokens > 0
```

Comment:
I also fixed this (for the above decoder tokens) in the prefix caching PR. Also to clarify the semantic of `num_new_tokens`:
- Before calling `_schedule_encoder_inputs`, `num_new_tokens` would be the text tokens as well as image tokens (placeholder).
- After calling `_schedule_encoder_inputs`, `num_new_tokens` may be the same as before if encoder budget allows; otherwise it would be reduced to only include text tokens.

Is this understanding correct?

Reply:
@comaniac Yes, correct. When the encoder cache or budget is insufficient, `num_new_tokens` can decrease up to the point just before the encoder input (e.g., image placeholder).